{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5c5a9f6-c13a-4b5c-bf43-61c0fe066082",
   "metadata": {},
   "source": [
    "# Projet de Génomique : Mapping \n",
    "### Fait par EL HOUMA Mohamed, DEKARI Adil et BIYIK Kenan "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f98087c-7d86-4efc-ba71-5188a102a65e",
   "metadata": {},
   "source": [
    "Durant ce projet, l'objectif était de développer une stratégie de << mapping >> de séquences \n",
    "génomiques courtes, appelées “reads” (ou lectures), sur un génome de référence donné."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa1b90b-f7b6-456a-9e32-def0d9a30108",
   "metadata": {},
   "source": [
    "Nous allons donc vous présenter de manière détaillée les différentes étapes d'élaboration de ce projet."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "080e963e-2541-4dcc-8e24-76ee815b8bd8",
   "metadata": {},
   "source": [
    "## I - Initialisation des données"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3789ce34-587a-4dd0-87ae-747f56d26892",
   "metadata": {},
   "source": [
    "Le jeu de données qui nous a été fournis était sous le format FastQ. Notre première mission était donc de récupérer notre génome et de le formater de façon à le rendre utilisable dans la suite de notre projet.\n",
    "Pour cela, nous nous sommes servi de la bibliothèque Biopython qui présente de nombreuses fonctions qui vont nous faciliter la manipulation de notre jeu de données."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02452410-a3e9-4e07-a3c4-fa2d7d8f6d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61eb0c4-69d9-478d-8133-937fab690c20",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from pathlib import Path\n",
    "import gzip, zipfile\n",
    "import time\n",
    "\n",
    "def genome(fasta_file):\n",
    "        sequences = []\n",
    "        open_func = gzip.open if fasta_file.endswith(\".gz\") else open\n",
    "        \n",
    "        with open_func(fasta_file, \"rt\") as handle:\n",
    "            for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                sequences += str(record.seq)\n",
    "\n",
    "        T = \"\".join(sequences).upper()\n",
    "\n",
    "        return T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57860947-2f48-4c5c-a907-54d7a50606a4",
   "metadata": {},
   "source": [
    "Ici, nous importons les bibliothèques nécessaires et définissons une fonction read_genome\n",
    "\n",
    "Cette fonction lira notre fichier fasta et le formatera en une unique chaîne de caractères que nous pourrons ensuite indexer avec DC3."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8988e2f5-de2b-4e3b-a794-cb55ecacb741",
   "metadata": {},
   "source": [
    "Maintenant que nos données sont utilisables, passons à la suite de ce projet."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc9b481-1c77-4142-acd4-eefb7f77544a",
   "metadata": {},
   "source": [
    "## II - Création d'une suffix array"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d2f1ba0-ce3c-4d1e-bf31-4404c915e59e",
   "metadata": {},
   "source": [
    "Dans cette partie, notre but est de définir une structure de données permettant de \n",
    "chercher des mots de longueur fixe à partir d’un texte d'une certaines longueurs.\n",
    "Ici, on va former ce qu'on appelle une << suffix array >>. \n",
    "Pour faire simple, c'est un tableau contenant les positions de tous les suffixes de la chaîne, triés par ordre lexicographique.\n",
    "Cette structure de données va nous permettre de rechercher de manière assez facile les reads découpés en k-mers sur notre génome de référence.\n",
    "Pour créer ce suffix array, on va se servir de l'algorithme DC3 dont vous trouverez le programme ci-dessous : "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3252c29c-76e5-4d44-8c42-fe7cf7d5b68c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# L'algorithme DC3 pour la construction du Suffix Array est en O(n) \n",
    "\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "start_time = time.perf_counter()\n",
    "\n",
    "def print_runtime():\n",
    "    elapsed = time.perf_counter() - start_time\n",
    "    return elapsed\n",
    "\n",
    "def presence_doublons(l): #pour la vérification de doublons\n",
    "    return len(l) != len(set(l))\n",
    "\n",
    "def radix_sort(L, num_keys):\n",
    "    if type(L[0][0])==str:\n",
    "        for i in range(len(L)):\n",
    "            col = []\n",
    "            for x in L[i]:\n",
    "                if type(x)==str:\n",
    "                    L[i] += [ord(x)]\n",
    "                else:\n",
    "                    L[i] += [x]\n",
    "\n",
    "    for key_index in range(num_keys - 1, -1, -1):\n",
    "        max_val = max(x[key_index] for x in L)\n",
    "        count = [0] * (max_val + 2)\n",
    "        for x in L:\n",
    "            count[x[key_index] + 1] += 1\n",
    "        for i in range(1, len(count)):\n",
    "            count[i] += count[i - 1]\n",
    "        res = [0] * len(L)\n",
    "        for x in L:\n",
    "            res[count[x[key_index]]] = x\n",
    "            count[x[key_index]] += 1\n",
    "\n",
    "        L = res\n",
    "    return L\n",
    "\n",
    "def merge(index_0, index_12, liste_T, iterations, rank_12):\n",
    "    index_012=[]\n",
    "    i_a=0\n",
    "    i_b=0\n",
    "\n",
    "    while i_a<len(index_0) and i_b<len(index_12):\n",
    "\n",
    "        a = index_0[i_a]\n",
    "        b = index_12[i_b]\n",
    "\n",
    "\n",
    "        if liste_T[iterations-1][a]<liste_T[iterations-1][b]:\n",
    "            index_012+=[a]\n",
    "            i_a+=1\n",
    "\n",
    "\n",
    "        elif liste_T[iterations-1][a]>liste_T[iterations-1][b]:\n",
    "            index_012+=[b]\n",
    "            i_b+=1\n",
    "\n",
    "\n",
    "        else :\n",
    "            if b%3==1 :\n",
    "                if rank_12.get(a+1, -1) < rank_12.get(b+1, -1):\n",
    "                    index_012+=[a]\n",
    "                    i_a+=1\n",
    "\n",
    "                else :\n",
    "                    index_012+=[b]\n",
    "                    i_b+=1\n",
    "\n",
    "            else :\n",
    "                if liste_T[iterations-1][a+1]<liste_T[iterations-1][b+1]:\n",
    "                    index_012+=[a]\n",
    "                    i_a+=1\n",
    "\n",
    "\n",
    "                elif liste_T[iterations-1][a+1]>liste_T[iterations-1][b+1]:\n",
    "                    index_012+=[b]\n",
    "                    i_b+=1\n",
    "\n",
    "\n",
    "                else:\n",
    "                    if rank_12.get(a+2, -1) < rank_12.get(b+2, -1):\n",
    "                        index_012+=[a]\n",
    "                        i_a+=1\n",
    "\n",
    "                    else :\n",
    "                        index_012+=[b]\n",
    "                        i_b+=1\n",
    "\n",
    "    if i_a==len(index_0):\n",
    "        index_012+=index_12[i_b:]\n",
    "    else:\n",
    "        index_012+=index_0[i_a:]\n",
    "\n",
    "    index_012.pop(0)\n",
    "\n",
    "    return(index_012)\n",
    "\n",
    "def DC3(T):\n",
    "    ## ETAPE 1 :\n",
    "    verif=True\n",
    "    iterations=0\n",
    "    liste_T=[T]\n",
    "    liste_P12=[]\n",
    "    liste_P0=[]\n",
    "    liste_P1=[]\n",
    "    \n",
    "    while verif:\n",
    "        if type(T[0])==str:\n",
    "            T+=['0','0','0']\n",
    "        else :\n",
    "            T+=[0,0,0]\n",
    "        \n",
    "        P0=[i for i in range (0,len(T)-2,3)]\n",
    "        P1=[i for i in range (1,len(T)-2,3)]\n",
    "        P2=[i for i in range (2,len(T)-2,3)]\n",
    "        P12=P1+P2\n",
    "        liste_P12+=[P12]\n",
    "        liste_P0+=[P0]\n",
    "        liste_P1+=[P1]\n",
    "        \n",
    "        if type(T[0])==str:\n",
    "            base_map = {'A': 1, 'C': 2, 'G': 3, 'T': 4, '0': 0}\n",
    "            T_array = np.array([base_map[b.upper()] for b in T])\n",
    "        else:\n",
    "            T_array = np.asarray(T)\n",
    "        P12_array = np.asarray(P12)\n",
    "        n = len(P12)\n",
    "\n",
    "        R12 = np.empty((n, 5), dtype=np.int64)\n",
    "        R12[:, 0] = T_array[P12_array]\n",
    "        R12[:, 1] = T_array[P12_array + 1]\n",
    "        R12[:, 2] = T_array[P12_array + 2]\n",
    "        R12[:, 3] = P12_array\n",
    "        R12[:, 4] = np.arange(n)\n",
    "        R12=R12.tolist()\n",
    "        \n",
    "        R12_sorted=radix_sort(R12,3)\n",
    "        \n",
    "        index_12=[R12_sorted[i][3] for i in range (len(R12_sorted))]\n",
    "\n",
    "        ordre=[1]\n",
    "        c=1\n",
    "        for i in range (1,len(R12_sorted)):\n",
    "            if R12_sorted[i][:3]!=R12_sorted[i-1][:3]:\n",
    "                c+=1\n",
    "            ordre+=[c]\n",
    "        \n",
    "        T_prime=[0]*len(P12)\n",
    "        for i in range (len(R12_sorted)):\n",
    "            T_prime[R12_sorted[i][4]]=ordre[i]\n",
    "        T=T_prime\n",
    "        liste_T+=[T]\n",
    "        iterations+=1\n",
    "        verif = presence_doublons(T)\n",
    "\n",
    "    while iterations!=1:\n",
    "        ## ETAPE 2 :\n",
    "        rank_12 = {val: i for i, val in enumerate(index_12)}\n",
    "\n",
    "        R0 = []\n",
    "        for i in range(len(liste_P1[iterations-1])):\n",
    "            var_1 = liste_T[iterations-1][liste_P0[iterations-1][i]]\n",
    "            var_2 = rank_12.get(liste_P1[iterations-1][i], -1) + 1\n",
    "            R0.append([var_1, var_2, liste_P0[iterations-1][i]])\n",
    "\n",
    "        if len(liste_P1[iterations-1])<len(liste_P0[iterations-1]):\n",
    "            R0+=[[liste_T[iterations-1][liste_P0[iterations-1][-1]],0,liste_P0[iterations-1][i]]]\n",
    "        \n",
    "        R0_sorted=radix_sort(R0,2)\n",
    "        \n",
    "        index_0 = [r[2] for r in R0_sorted]\n",
    "        ordre_0 = [i for i in range (1,len(index_0)+1)]\n",
    "        \n",
    "        ## ETAPE 3 :\n",
    "        rank_12 = {val: pos for pos, val in enumerate(index_12)}\n",
    "\n",
    "        index_012 = merge(index_0, index_12, liste_T, iterations, rank_12)\n",
    "        iterations-=1\n",
    "\n",
    "        P12_n=liste_P12[iterations-1]\n",
    "        index_12=[P12_n[i] for i in index_012]\n",
    "\n",
    "    end_time=print_runtime()\n",
    "    \n",
    "    return index_12, end_time\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba4f56-a64e-4981-b3e1-82ca71b26761",
   "metadata": {},
   "source": [
    "Notre suffix array est fin prêt. On peut donc passer à l'étape suivante."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f54778e4-f04e-4d60-a3a4-575f63eefa12",
   "metadata": {},
   "source": [
    "## III - Découpage des lectures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54277e34-3671-46ae-b9e1-d43554589010",
   "metadata": {},
   "source": [
    "Pour faire le mapping de nos lectures de manière précise, on va les découper en k-mers dont on va définir la taille (celle-ci va être choisis selon différents facteurs, essentiellement le temps de recherche, la mémoire utilisée et l'ambiguité du mapping)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27de1dc1-5931-4233-8c01-bc319fc023dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def k_mers(seq, k):\n",
    "    return [seq[i:i+k] for i in range(len(seq) - k + 1)]\n",
    "\n",
    "read = \"AGCTTAGC\"\n",
    "k = 3\n",
    "print(k_mers(read, k))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e057d1b-c557-4038-bd84-a109bdec302b",
   "metadata": {},
   "source": [
    "## IV - Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afdaa35d-8a9b-497c-8239-558fb6d3ae65",
   "metadata": {},
   "source": [
    "Notre structure étant créer, on peut donc passer au mapping, c'est-à-dire à la recherche des k-mers dans le suffix array qu'on a créé précedemment.\n",
    "Pour faire cela, on va utiliser ce qu'on appelle la Burrows-Wheeler Transform (BWT)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4385dbef-4acb-44c7-8805-b64c3ffb434b",
   "metadata": {},
   "source": [
    "Dans un premier temps, on fabrique tous nos circular shifts :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6046c09f-42ac-46f5-a817-d066bcde9147",
   "metadata": {},
   "outputs": [],
   "source": [
    "def circular_shift(text,k):\n",
    "    return text[-k:]+text[:-k]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4c54050-347c-4297-9bb2-393b9d96f489",
   "metadata": {},
   "source": [
    "Ensuite, on applique la BWT : "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2fd3b6e-4644-4912-91f9-58133a276f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def BWT(text, suffix_array):\n",
    "    bwt = [''] * len(suffix_array)\n",
    "    for i in range(len(suffix_array)):\n",
    "        pos = suffix_array[i]\n",
    "        if pos == 0:\n",
    "            bwt[i] = text[-1] \n",
    "        else:\n",
    "            bwt[i] = text[pos - 1]\n",
    "    return \"\".join(bwt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001e0538-f7d9-44f3-b4ea-95499a73f6bf",
   "metadata": {},
   "source": [
    "Inversion de la séquence pour rechercher dans la séquence complémentaire: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e506e569-1e16-408c-b3f0-cce500422b37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complémentaire(seq):\n",
    "    \"\"\"Donne le brin complémentaire inversé.\"\"\"\n",
    "    a_inverser = str.maketrans(\"ACGTN$\", \"TGCAN$\", \" \") \n",
    "    return seq.translate(a_inverser)[::-1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f17049-cb03-443b-968b-3b8ba483157c",
   "metadata": {},
   "source": [
    "Recherche avec l'algorithme de backward search (c correspond à la position, Occ au nombre): "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "238e5e3a-0a51-4547-86de-4848c65aa6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Définit la fonction pour créer la table C \n",
    "def construire_C(bwt):\n",
    "    counts = defaultdict(int)\n",
    "    # Parcourt chaque caractère dans la BWT \n",
    "    for c in bwt:\n",
    "        counts[c] += 1\n",
    "    # Récupère tous les caractères uniques et les trie par ordre lexicographique\n",
    "    chars = sorted(counts.keys())\n",
    "    # Initialise le dictionnaire C final\n",
    "    C = {}\n",
    "    total = 0\n",
    "    for c in chars:\n",
    "        C[c] = total\n",
    "        # Ajoute le nombre d'occurrences de c au total pour la prochaine itération\n",
    "        total += counts[c]\n",
    "    # Renvoie la table C complétée\n",
    "    return C\n",
    "\n",
    "# Définit la fonction pour créer la table Occ (occurrences)\n",
    "def construire_Occ(bwt):\n",
    "    # Récupère l'alphabet de la BWT\n",
    "    chars = sorted(set(bwt))\n",
    "    occ = {c: [0] for c in chars}\n",
    "    for ch in bwt:\n",
    "        for c in chars:\n",
    "            occ[c].append(occ[c][-1] + (1 if c == ch else 0))\n",
    "    return occ\n",
    "\n",
    "# Définit la fonction d'accès rapide à la table Occ (combien de c avant la position)\n",
    "def Occ(occ, c, pos):\n",
    "    if pos < 0:\n",
    "        return 0\n",
    "    # Renvoie le compte précalculé à la position pos \n",
    "    return occ[c][pos+1]\n",
    "\n",
    "# Définit la fonction de recherche arrière (Backward Search)\n",
    "def backward_search(pattern, bwt, suffix_array, C, Occ_table):  \n",
    "    # Initialise le pointeur de début de plage (indice Suffix Array)\n",
    "    start = 0\n",
    "    # Initialise le pointeur de fin de plage (indice Suffix Array)\n",
    "    end = len(bwt) - 1\n",
    "\n",
    "    # Parcourt le pattern (le motif) à l'envers, du dernier caractère au premier\n",
    "    for i in range(len(pattern)-1, -1, -1):\n",
    "        # Récupère le caractère actuel du motif\n",
    "        c = pattern[i]\n",
    "        # Si ce caractère n'est pas dans le génome\n",
    "        if c not in C:\n",
    "            # le motif n'existe pas, on renvoie 0 occurrence et une liste vide\n",
    "            return 0, []  \n",
    "        \n",
    "        # Calcule la nouvelle position de début\n",
    "        pos_start = C[c] + Occ(Occ_table, c, start - 1)\n",
    "        # Calcule la nouvelle position de fin\n",
    "        pos_end = C[c] + Occ(Occ_table, c, end) - 1\n",
    "        \n",
    "        start = pos_start\n",
    "        end = pos_end\n",
    "\n",
    "        if start > end:\n",
    "            #le motif n'existe pas on arrête\n",
    "            return 0, []\n",
    "\n",
    "    # Calcule le nombre total d'occurrences\n",
    "    count = end - start + 1\n",
    "    # Récupère les positions de départ réelles dans le génome en consultant le Suffix Array\n",
    "    positions = [suffix_array[i] for i in range(start, end + 1)]\n",
    "    # Renvoie le nombre d'occurrences et la liste des positions\n",
    "    return count, positions\n",
    "\n",
    "# Définit la fonction pour extraire un k-mer à une position i\n",
    "def get_kmer_at_position(genome_text, i, k):\n",
    "    \n",
    "    start_index = i - 1\n",
    "    \n",
    "    # Calcule l'index de fin\n",
    "    end_index = start_index + k\n",
    "    \n",
    "    # Extrait et renvoie la sous-chaîne (le k-mer)\n",
    "    return genome_text[start_index : end_index]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8675bda-46e8-4849-9eca-83d9987869f6",
   "metadata": {},
   "source": [
    "Pour le brin complémentaire : soit on applique DC3 2x pour le brin sens et le brin antisens, soit on colle le brin sens et le brin antisens pour avoir un grand génome et appliquait la recherche dessus.\n",
    "Pas besoin de prendre le complémentaire de la BWT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0420a4-5697-4b31-9b99-c8c80deec104",
   "metadata": {},
   "source": [
    "Voir BWT non naïve pour le deuxième tiré de la mise en oeuvre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bd6bdc6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 0. Paramètres\n",
    "# On définit ici la taille du k-mer\n",
    "taille_kmer = 25\n",
    "\n",
    "# Importation des fichiers (A changer pour l'utilisateur)\n",
    "fasta_file = r\"C:\\\\Users\\\\melho\\\\Downloads\\\\GCF_000002765.5_GCA_000002765_genomic (1).fna.gz\"\n",
    "fastq_file = r\"C:\\\\Users\\\\melho\\\\Downloads\\\\single_Pfal_dat (1).fq.gz\"\n",
    "\n",
    "\n",
    "# 1. Charger le génome ---\n",
    "T = genome(fasta_file) \n",
    "\n",
    "# 2. Construire le Suffix Array (DC3)\n",
    "suffix_array, runtime = DC3(list(T))\n",
    "print(f\"DC3 a pris {runtime} secondes.\")\n",
    "print(f\"la suffix array est :\")\n",
    "print(suffix_array[:100])\n",
    "print('')\n",
    "\n",
    "# 3. Construire la BWT \n",
    "bwt = BWT(T, suffix_array) \n",
    "C_table = construire_C(bwt)\n",
    "Occ_table = construire_Occ(bwt)\n",
    "    \n",
    "reads_trouves = 0\n",
    "total_reads = 0\n",
    "    \n",
    "# On prend chaque read de notre fichier fastq_file et le recherche dans notre génome de référence\n",
    "\n",
    "# 4. Boucle de Mapping ---\n",
    "with gzip.open(fastq_file, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fastq\"):\n",
    "        total_reads += 1\n",
    "        read_id = record.id\n",
    "        read_seq = str(record.seq)\n",
    "            \n",
    "        # On ne prend que le premier k-mer comme \"seed\"\n",
    "        seed = read_seq[:taille_kmer]\n",
    "            \n",
    "        # 6. Recherche du k-mer sur le brin sens (+)\n",
    "        count_fwd, positions_fwd = backward_search(seed, bwt, suffix_array, C_table, Occ_table)\n",
    "            \n",
    "        # 7. Recherche du k-mer sur le brin complémentaire (-)\n",
    "        seed_rev = complémentaire(seed)\n",
    "        count_rev, positions_rev = backward_search(seed_rev, bwt, suffix_array, C_table, Occ_table)\n",
    "            \n",
    "#       8. Affichage des résultats\n",
    "        if count_fwd > 0:\n",
    "            reads_trouves += 1\n",
    "            print(f\"Read '{read_id}' TROUVÉ (Brin '+') : {count_fwd} aux positions {positions_fwd}\")\n",
    "            \n",
    "        if count_rev > 0:\n",
    "            reads_trouves += 1\n",
    "            print(f\"Read '{read_id}' TROUVÉ (Brin '-') : {count_rev} aux positions {positions_rev}\")\n",
    "\n",
    "    print(f\"Mapping terminé\")\n",
    "    print(f\"Reads analysés : {total_reads}\")\n",
    "    print(f\"Reads mappés : {reads_trouves}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8551864-4e80-40d7-8c15-94fe31f4146a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Un extrait de la séquence test :\n",
    "print(T[:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed21cd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Si l'on cherche un motif spécifique dans le génome     \n",
    "# Spécifiez la séquence à rechercher\n",
    "mon_pattern = list(input(\"séquences à rechercher : \"))\n",
    "print(mon_pattern)\n",
    "count, positions = backward_search(mon_pattern, bwt, suffix_array, C_table, Occ_table)\n",
    "\n",
    "if count > 0:\n",
    "    print(f\"Motif trouvé {count} fois et aux positions {positions}\")\n",
    "else:\n",
    "    print(\"Motif non trouvé.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41539b42",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Si l'on cherche à extraire un k-mer à une position donnée\n",
    "#  Spécifiez la position et la longueur\n",
    " \n",
    "position_i = int(input(\"Entrez la position i : \")) # i >= 1\n",
    "longueur_k = int(input(\"Entrez la longueur k : \")) # k > 0 \n",
    "\n",
    "séquence_rechercher = get_kmer_at_position(T, position_i, longueur_k)\n",
    "\n",
    "print(f\"La séquence recherchée est : {séquence_rechercher}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
